tilequeue/process.py

```diff
@@ -580,3 +582,191 @@ def convert_source_data_to_feature_layers(rows, layer_data, bounds, zoom):
+
+
+def calculate_cut_coords_by_zoom(
+        coord, metatile_zoom, cfg_tile_sizes, max_zoom):
+    """
+    Returns a map of nominal zoom to the list of cut coordinates at that
+    nominal zoom.
+    """
+
+    tile_sizes_by_zoom = calculate_sizes_by_zoom(
+        coord, metatile_zoom, cfg_tile_sizes, max_zoom)
+
+    cut_coords_by_zoom = {}
+    for nominal_zoom, tile_sizes in tile_sizes_by_zoom.iteritems():
+        cut_coords = []
+        for tile_size in tile_sizes:
+            cut_coords.extend(metatile_children_with_size(
+                coord, metatile_zoom, nominal_zoom, tile_size))
+
+        cut_coords_by_zoom[nominal_zoom] = cut_coords
+
+    return cut_coords_by_zoom
+
+
+class Processor(object):
+    def __init__(self, coord, metatile_zoom, fetch_fn, layer_data,
+                 post_process_data, formats, buffer_cfg, output_calc_mapping,
+                 max_zoom, cfg_tile_sizes):
+        self.coord = coord
+        self.metatile_zoom = metatile_zoom
+        self.fetch_fn = fetch_fn
+        self.layer_data = layer_data
+        self.post_process_data = post_process_data
+        self.formats = formats
+        self.buffer_cfg = buffer_cfg
+        self.output_calc_mapping = output_calc_mapping
+        self.max_zoom = max_zoom
+        self.cfg_tile_sizes = cfg_tile_sizes
+
+    def fetch(self):
+        unpadded_bounds = coord_to_mercator_bounds(self.coord)
+
+        cut_coords_by_zoom = calculate_cut_coords_by_zoom(
+            self.coord, self.metatile_zoom, self.cfg_tile_sizes, self.max_zoom)
+        feature_layers_by_zoom = {}
+
+        for nominal_zoom, _ in cut_coords_by_zoom.items():
+            source_rows = self.fetch_fn(nominal_zoom, unpadded_bounds)
+            feature_layers = convert_source_data_to_feature_layers(
+                source_rows, self.layer_data, unpadded_bounds, self.coord.zoom)
+            feature_layers_by_zoom[nominal_zoom] = feature_layers
+
+        self.cut_coords_by_zoom = cut_coords_by_zoom
+        self.feature_layers_by_zoom = feature_layers_by_zoom
+
+    def process_tiles(self):
+        unpadded_bounds = coord_to_mercator_bounds(self.coord)
+
+        all_formatted_tiles = []
+        all_extra_data = {}
+
+        for nominal_zoom, cut_coords in self.cut_coords_by_zoom.items():
+            feature_layers = self.feature_layers_by_zoom[nominal_zoom]
+            formatted_tiles, extra_data = process_coord(
+                self.coord, nominal_zoom, feature_layers,
+                self.post_process_data, self.formats, unpadded_bounds,
+                cut_coords, self.buffer_cfg, self.output_calc_mapping
+            )
+            all_formatted_tiles.extend(formatted_tiles)
+            all_extra_data.update(extra_data)
+
+        return all_formatted_tiles, all_extra_data
+
+
+def process(coord, metatile_zoom, fetch_fn, layer_data, post_process_data,
+            formats, buffer_cfg, output_calc_mapping, max_zoom,
+            cfg_tile_sizes):
```

Comment:
I went backwards and forwards on the interface. Overall, I think a single function call is cleaner, but I had trouble separating out the exceptions in the fetch phase from the processing phase. That's a helpful distinction, so I thought it would be good to keep.

Anyone know how to cleanly support `try:/except:` changing or wrapping the exception while keeping the stack trace intact? I feel like this is something I used to know, but have forgotten :confused:

Reply:
Sorry, I just meant that we can keep the single function call, just have most of the "config" state in the closure/instance.